src/EXTRA-FIX/fix_wall_flow.cpp

```diff
@@ -0,0 +1,294 @@
+          return domain->lattice->zlattice;
+        default: return 0.0;
+      }
+      return 0.0;
+  };
+  double scale = getscale();
+
+  if (narg - iarg == wallcount + 2)
+  {
+    if(strcmp(arg[narg - 2], "units") != 0) error->all(FLERR, "Wrong fix wall/flow units command");
+    if (strcmp(arg[narg - 1], "box") == 0) scale = 1.0;
+    else if (strcmp(arg[narg - 1], "lattice") == 0)
+    {
+      scale = getscale();
+    }
+    else error->all(FLERR, "Wrong fix wall/flow units command");
+  }
+
+  walls.resize(wallcount + 2);
+  walls.front() = domain->boxlo[flowax];
+  for (size_t w = 1; w <= wallcount; ++w, ++iarg)
+  {
+    walls[w] = utils::numeric(FLERR,arg[iarg],do_abort,lmp) * scale;
+  }
+  walls.back() = domain->boxhi[flowax];
+  if (!std::is_sorted(walls.begin(), walls.end(), std::less_equal<double>()))
+  {
+    error->all(FLERR, "Wrong fix wall/flow wall ordering or some walls are outside simulation domain");
+  }
+  std::cout << "Walls:\n";
+  for (auto w : walls)
+  {
+    std::cout << w << " ";
+  }
+  std::cout << std::endl;
+
+  memory->grow(current_segment, atom->nmax, "WallFlow::current_segment");
+  atom->add_callback(Atom::GROW);
+  if (restart_peratom) atom->add_callback(Atom::RESTART);
+
+  maxexchange = 1;
+
+  random = new RanMars(lmp, rndseed + comm->me);
+ }
+
+/* ---------------------------------------------------------------------- */
+
+FixWallFlow::~FixWallFlow()
+{
+  if (copymode) return;
+  atom->delete_callback(id, Atom::GROW);
+  if (restart_peratom) atom->delete_callback(id, Atom::RESTART);
+  memory->destroy(current_segment);
+
+  delete random;
+}
+
+/* ---------------------------------------------------------------------- */
+
+int FixWallFlow::setmask()
+{
+  int mask = 0;
+
+  mask |= END_OF_STEP;
+
+  return mask;
+}
+
+/* ---------------------------------------------------------------------- */
+
+void FixWallFlow::init()
+{
+  int nrigid = 0;
+  for (int i = 0; i < modify->nfix; i++)
+    if (modify->fix[i]->rigid_flag) nrigid++;
+
+  if (nrigid && comm->me == 0)
+    error->warning(FLERR,"FixWallFlow is not compatible with rigid bodies");
+
+  for (int i = 0; i < atom->nlocal; ++i)
+  {
+    double pos = atom->x[i][flowax];
+    current_segment[i] = compute_current_segment(pos);
+  }
+}
+
+/* ---------------------------------------------------------------------- */
+
+void FixWallFlow::end_of_step()
+{
+  double **x = atom->x;
+  int *mask = atom->mask;
+  int nlocal = atom->nlocal;
+
+  for (int i = 0; i < nlocal; ++i)
+  {
+    if (mask[i] & groupbit)
+    {
+       double pos = x[i][flowax];
+       int prev_segment = current_segment[i];
+       current_segment[i] = compute_current_segment(pos);
+
+       if (prev_segment != current_segment[i])
+       {
+          generate_velocity(i);
+       }
+    }
+  }
+}
+
+void FixWallFlow::generate_velocity(int atom_i)
+{
+  const int newton_iteration_count = 10;
+  double *vel = atom->v[atom_i];
+  double mass = atom->mass[atom->type[atom_i]];
```

Comment:
This assumes that masses are per-type, but that may not be true. For systems with per-atom masses, you must use `atom->rmass[atom_i]`. You would get an incorrect mass or a crash due to a segmentation fault otherwise. See, for instance, fix nve.

Reply:
I took this into account in the kokkos version, but somehow forgot it here, fixed